File: rust/playground/src/doublylinkedlist.py
class Node:
    def __init__(self,
                key: int,
                val: int,
                next: 'Node | None' = None,
                prev: 'Node | None' = None
                ) -> None:
        self.key = key
        self.val = val
        self.next = next
        self.prev = prev


class DoublyLinkedList:
    def __init__(self):
        self.size = 0
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next, self.tail.prev = self.tail, self.head   # head <-> tail

    def push(self, node: Node): # O(1)
        head, nxt = self.head, self.head.next   # head <-> nxt <-> ... <-> tail
        head.next, nxt.prev = node, node        # head  -> node <-  nxt     #type: ignore
        node.prev, node.next = head, nxt        # head <-> node <-> nxt
        self.size += 1

    def remove(self, node: Node) -> None: # O(1)
        # head <-> ... <-> node <-> ... <-> tail
        prev, nxt = node.prev, node.next        # prev <-> node <-> nxt
        node.prev, node.next = None, None       # prev  -> node <-  nxt
        prev.next, nxt.prev = nxt, prev         # prev <-> nxt   |  node    #type: ignore
        self.size -= 1

    def pop(self) -> Node: # O(1)
        node = self.tail.prev                   # head <-> ... <-> node <-> tail
        self.remove(node)       # type: ignore
        return node             # type: ignore

File: rust/playground/src/doubly_linked_list.rs
#![allow(unused)]

use std::cell::RefCell;
use std::rc::Rc;

struct Node {
    val: i32,
    prev: Option<Rc<Node>>,
    next: Option<Rc<Node>>,
}

fn main() {
    //  1 <-> 2
    let mut n1 = Rc::new(Node {
        val: 1,
        prev: None,
        next: None,
        // next: Some(Rc::clone(&n2)),
    });

    let mut n2 = Rc::new(Node {
        val: 1,
        prev: Some(Rc::clone(&n1)),
        next: None,
    });

    // 1 -> 2
    n1.next = Some(Rc::clone(&n2));

    // 1 <-> 2
    // n2.prev = Some(Rc::clone(&n1));
}

fn print_list(head: Node) {
    let curr = &head;

    // while let node = *curr {
    //     println!("{}", node.val);
    // }
}

/*
struct DoublyLinkedList {
    size: u32,
    head: Node,
    tail: Node,
}

impl DoublyLinkedList {
    fn new() -> Self {
        let mut head = Node {
            // key: -1,
            val: -1,
            next: &Rc::,
            prev: None,
        };
        let mut tail = Node {
            // key: -1,
            val: -1,
            next: None,
            prev: None,
        };
        head.next = Some(Rc::new(tail));

        tail.prev = Some(Rc::clone(head));

        Self {
            size: 0,
            head: Node {
                val: -1,
                next: None,
                prev: None,
            },
            tail: Node {
                val: -1,
                next: None,
                prev: None,
            },
        }
    }
}

struct LRUCache {}

impl LRUCache {
    fn new(capacity: i32) -> Self {
        LRUCache {}
    }
    fn get(&self, key: i32) -> i32 {
        1
    }
    fn put(&self, key: i32, value: i32) {}
}

#[test]
fn test_1() {
    let _ = LRUCache::new(10);
}
*/
